### PR TITLE
Use the same criteria to flip chiral atoms in SMILES parsing/writing (Fixes #1028)

### LINKED ISSUE #1028: Alternating canonical SMILES for fused ring with N
reported by @adalke as part of the discussion of #1022 .
Here's a reproducible (adapted from Andrew's):

```
In [1]: from rdkit import Chem

In [2]: s = 'O[C@H]1CC2CCC(C1)[N@@]2C'

In [3]: m = Chem.MolFromSmiles(s)

In [4]: Chem.MolToSmiles(m,isomericSmiles=True)
Out[4]: 'C[N@]1C2CCC1C[C@@H](O)C2'

In [5]: Chem.MolToSmiles(Chem.MolFromSmiles(_4),isomericSmiles=True)
Out[5]: 'C[N@]1C2CCC1C[C@H](O)C2'

In [6]: Chem.MolToSmiles(Chem.MolFromSmiles(_5),isomericSmiles=True)
Out[6]: 'C[N@]1C2CCC1C[C@@H](O)C2'
```

@NadineSchneider : This is a canonicalization failure, but it may not be completely due to your code. Do you have time to take a look?

### PR DESCRIPTION
**Pull request by ricrogz**
Explanation of the problem is in the comments of Issue #1028 . But it summarizes into that SMILES parsing and writing were using different criteria to decide if the chiral tag of an atom had to be flipped or not.

This PR moves the code to decide whether to flip a chiral atom into Canon.cpp, and uses the same function to check chiral atoms at writing and parsing of SMILES strings.

I have not checked the change in the python test, but both strings seem to be "stable" and not flip on roundtripping the SMILES -- if the fix of the reference is wrong, then maybe the algorithm to decide the flipping is not quite correct.

### CODE REVIEW

**Comment by greglandrum** [comment]
Thanks for digging into this and fixing a bug I've been avoiding for far too long. :-)

Just for completeness, please add a test case for the example from #1028 as well: `O[C@H]1CC2CCC(C1)[N@@]2C`

File: Code/GraphMol/Wrap/rough_test.py [hunk 1]
```diff
@@ -3771,7 +3771,7 @@ def testSequenceBasics(self):
     helm = 'PEPTIDE1{C.Y.I.Q.N.C.P.L.G}$$$$'
     seq = 'CYIQNCPLG'
     fasta = '>\nCYIQNCPLG\n'
-    smi = 'CC[C@H](C)[C@H](NC(=O)[C@H](Cc1ccc(O)cc1)NC(=O)[C@@H](N)CS)C(=O)N[C@@H](CCC(N)=O)C(=O)N[C@@H](CC(N)=O)C(=O)N[C@@H](CS)C(=O)N1CCC[C@H]1C(=O)N[C@@H](CC(C)C)C(=O)NCC(=O)O'
+    smi = 'CC[C@H](C)[C@H](NC(=O)[C@H](Cc1ccc(O)cc1)NC(=O)[C@@H](N)CS)C(=O)N[C@@H](CCC(N)=O)C(=O)N[C@@H](CC(N)=O)C(=O)N[C@@H](CS)C(=O)N1CCC[C@@H]1C(=O)N[C@@H](CC(C)C)C(=O)NCC(=O)O'
```
Comment (greglandrum) [inline Code/GraphMol/Wrap/rough_test.py]: I'm pretty sure that the original SMILES here was correct and that you've inverted the stereochemistry of the proline.

**Comment by ricrogz** [comment]
> Thanks for digging into this and fixing a bug I've been avoiding for far too long. :-)
> 
> Just for completeness, please add a test case for the example from #1028 as well: `O[C@H]1CC2CCC(C1)[N@@]2C`

No problem, Greg. It was quite entertaining to look into this :)

I added the molecule from the issue, and fixed the issue with the sequence by also considering the implicit value when making the decision to flip a chiral center.

File: Code/GraphMol/Wrap/rough_test.py (hunk 1 above)
Comment (ricrogz) [reply Code/GraphMol/Wrap/rough_test.py]: You are right, it was ok. Apparently, at the time of deciding if the chiral center should be flipped, it did not have an explicit H, but an implicit one, and ended up being flipped.

I have added the implicit valence to the flipping decision, which seems to work fine, and solves this problem.

File: Code/GraphMol/Wrap/rough_test.py [hunk 2]
```diff
@@ -8,19 +8,25 @@
...
-
-from rdkit import __version__
-
+import doctest
+import gc
+import gzip
 # Boost functions are NOT found by doctest, this "fixes" them
 #  by adding the doctests to a fake module
 import importlib.util
+import os
```
Comment (greglandrum) [inline Code/GraphMol/Wrap/rough_test.py]: I understand why the automatic code formatter did this, but it really made the code less readable.
Please undo this change (so that the `import importlib.util` line and the comment before it are directly above line 30

Comment (ricrogz) [reply Code/GraphMol/Wrap/rough_test.py]: Oh! I'm sorry, I didn't even notice it did that. I just reverted it.
